replay/src/replay.ts

```diff
@@ -34,17 +85,159 @@ export class Replay extends Subprocess implements IReplayIpcNodeDelegate {
+      }
    }
 
    protected loop(): void {
       logger.info(`Replay ${this.id}: loop`);
-      this.stop(false); // just once for now
+
+      let finished = true;
+      const thisLoopStart = Date.now();
+      const nextLoopStart = thisLoopStart.valueOf() + this.interval;
+
+      // && this.queryInInterval(this.workloadIndex, nextLoopStart)
+      while (this.workloadIndex < this.workload!.length ) {
+
+         const delay = this.getDelayForIndex(this.workloadIndex, thisLoopStart);
+         const currentIndex = this.workloadIndex;
+         const currentQuery = this.workload![currentIndex];
+
+         setTimeout(() => {
+          this.processQuery(currentQuery.command_type, currentQuery.argument); }, delay);
+
+         // don't let the subprocess end because we still need to run these queries.
+         finished = false;
+         this.workloadIndex += 1;
+      }
+
+      const end = new Date();
+      const start = new Date(end.getTime() - this.interval - this.config.intervalOverlap);
+      if (start.getTime() < this.startTime!.getTime()) {
+         start.setTime(this.startTime!.getTime());
+      }
+      this.sendMetricsToS3(start, end);
+
+      if (finished) {
+        this.stop(false); // just once for now
+      }
    }
 
    protected async teardown(): Promise<void> {
+      logger.info(`Replay ${this.id}: teardown`);
+
+      if (this.error === false) {
+         await Replay.updateReplayStatus(this.id, ChildProgramStatus.DONE);
+      }
+
       this.ipcNode.stop();
    }
 
+   private async getWorkload() {
+
+      logger.info(`Getting workload from storage`);
+      this.workloadLocation = await Replay.getWorkloadForCapture(this.config.captureId);
+      this.workload = await this.storage.readJson(this.workloadLocation) as any;
+   }
+
+   // queryInInterval takes the index of the query in the workload and the time
+   //                 that next loop will begin in milliseconds and returns true
+   //                 if the query should be scheduled for the current loop otherwise false.
+   private queryInInterval(index: number, intervalStartTime: number): boolean {
+
+      const delay = this.getDelayForIndex(index, intervalStartTime);
+      return (delay >= 0 && delay < this.interval) ? true : false  ;
+   }
+
+   // getDelayForIndex takes in an index and the current interval's start time and returns the
+   //                  number of milliseconds to delay from the interval's start time.
+   private getDelayForIndex(index: number, intervalStartTime: number): number {
+      if (this.workload == null) {
+         return 0;
+      } else if (index > 0 && index < this.workload!.length) {
+         const queryStartTime: Date = new Date(this.workload![index].event_time);
+         const workloadStartTime: Date = new Date(this.workload![0].event_time);
+         const relativeQueryST = queryStartTime.getTime() - workloadStartTime.getTime() + this.startTime!.getTime();
```

Comment:
if it isn't going to take too long, I'd like to have the replay timing figured out before merging